File: Challenges/Python/ReverseLinkedListII.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def reverseBetween(self, head, start, stop):
        nodeBeforeReverse = None
        nodeAfterReverse = None
        reversedSectionHead = None
        reversedSectionEnd = None
        index = 1
        nodeBeforeReverseIndex = start - 1
        nodeAfterReverseIndex = stop + 1
        current = head
        reversedCurrent = None
        reversedPrevious = None
        while current is not None and index <= nodeAfterReverseIndex:
            if index >= nodeBeforeReverseIndex and index <= nodeAfterReverseIndex:
                if index == nodeBeforeReverseIndex:
                    nodeBeforeReverse = current
                elif index == nodeAfterReverseIndex:
                    nodeAfterReverse = current
                else:
                    reversedCurrent = ListNode(current.val)
                    if reversedPrevious:
                        reversedCurrent.next = reversedPrevious
                    if index == start:
                        reversedSectionEnd = reversedCurrent
                    if index == stop:
                        reversedSectionHead = reversedCurrent
                    reversedPrevious = reversedCurrent
            current = current.next
            index += 1
        if nodeBeforeReverse:
            nodeBeforeReverse.next = reversedSectionHead
        reversedSectionEnd.next = nodeAfterReverse
        return (reversedSectionHead, head)[nodeBeforeReverse is not None]

File: Challenges/JavaScript/ReverseLinkedListII.js
var getNode = function(val) {
   return {
      val: val,
      next: null
   };
};

var createListAndReturnHead = function(arr) {
   if (!arr || arr.length <= 0) {
      return { val: null, next: null };
   }
   const head = getNode(arr[0]);
   let prev = head;
   let current;
   for (let i = 1; i < arr.length; i++) {
      current = getNode(arr[i]);
      prev.next = current;
      prev = current;
   }

   return head;
};

var convertBackToArray = function(head) {
   let current = head;
   const array = [];

   while (current !== null) {
      array.push(current.val);
      current = current.next;
   }
   return array;
};

var reverseBetween = function(head, start, stop) {
   let nodeBeforeReverse = null;
   let nodeAfterReverse = null;
   let reversedSectionHead = null;
   let reversedSectionEnd = null;
   let index = 1;
   let nodeBeforeReverseIndex = start - 1;
   let nodeAfterReverseIndex = stop + 1;
   let current = head;
   let reversedCurrent = null;
   let reversedPrevious = null;

   while (current !== null && index <= nodeAfterReverseIndex) {
      if (index >= nodeBeforeReverseIndex && index <= nodeAfterReverseIndex) {
         if (index === nodeBeforeReverseIndex) nodeBeforeReverse = current;
         else if (index === nodeAfterReverseIndex) nodeAfterReverse = current;
         else {
            reversedCurrent = getNode(current.val);
            if (reversedPrevious) reversedCurrent.next = reversedPrevious;
            if (index === start) reversedSectionEnd = reversedCurrent;
            if (index === stop) reversedSectionHead = reversedCurrent;
            reversedPrevious = reversedCurrent;
         }
      }
      current = current.next;
      index++;
   }

   if (nodeBeforeReverse) nodeBeforeReverse.next = reversedSectionHead;
   reversedSectionEnd.next = nodeAfterReverse;
   return nodeBeforeReverse ? head : reversedSectionHead;
};

const head = createListAndReturnHead([1, 2, 3, 4, 5]);
const m = 2;
const n = 4;

const expectedOutput = [1, 4, 3, 2, 5];
const myOutput = convertBackToArray(reverseBetween(head, m, n));

console.log("expected ouput:", expectedOutput);
console.log("my output:     ", myOutput);
